proposals/flexible-vectors/README.md

```diff
@@ -1,7 +1,1184 @@
+- `vec.v64.lut2_m(idx: vec.v64, a: vec.v64, b: vec.v64, fallback: vec.v64) -> vec.v64`
+- `vec.v128.lut2_m(idx: vec.v128, a: vec.v128, b: vec.v128, fallback: vec.v128) -> vec.v128`
+
+```python
+def vec.S.lut2_m(idx, a, b, fallback):
+    result = vec.S.New()
+    for i in range(vec.S.length):
+        if idx[i] < vec.S.length:
+            result[i] = a[idx[i]]
+        elif idx[i] < 2*vec.S.length:
+            result[i] = b[idx[i] - vec.S.length]
+        else:
+            result[i] = fallback[i]
+    return result
+```
+
+### V128 shuffle
+
+Applies shuffle to each v128 of the vector.
+
+- `vec.i8x16.shuffle(a: vec.v128, b: vec.v128, imm: ImmLaneIdx32[16]) -> vec.v128`
+
+```python
+def vec.i8x16.shuffle(a, b, imm):
+    result = vec.v128.New()
+    for i in range(vec.v128.length):
+        result[i] = i8x16.shuffle(a[i], b[i], imm)
+    return result
+```
+
+### V128 swizzle
+
+Applies swizzle to each v128 of the vector.
+
+- `vec.i8x16.swizzle(a: vec.v128, s: vec.v128) -> vec.v128`
+
+```python
+def vec.i8x16.swizzle(idx, a, s):
+    result = vec.v128.New()
+    for i in range(vec.v128.length):
+        result[i] = i8x16.swizzle(a[i], s[i], imm)
+    return result
+```
+
+### Splat lane
+
+Gets a single lane from vector and broadcast it to the entire vector.
+`idx` is interpreted modulo the cardinal of the vector.
+
+- `vec.v8.splat_lane(v: vec.v8, idx: i32) -> vec.v8`
+- `vec.v16.splat_lane(v: vec.v16, idx: i32) -> vec.v16`
+- `vec.v32.splat_lane(v: vec.v32, idx: i32) -> vec.v32`
+- `vec.v64.splat_lane(v: vec.v64, idx: i32) -> vec.v64`
+- `vec.v128.splat_lane(v: vec.v128, idx: i32) -> vec.v128`
+
+```python
+def vec.S.splat_lane(v, imm):
+    idx = idx % vec.S.length
+    result = vec.S.New()
+    for i in range(vec.S.length):
+        result[i] = v[idx]
+    return result
+```
+
+### Concat
+
+Copies elements from vector `a` from first active element to last active element.
+Inner inactive elements are also copied.
+The remaining elements are set from the first elements from `b`.
+
+- `vec.v8.concat(m: vec.m8, a: vec.v8, b: vec.v8) -> vec.v8`
+- `vec.v16.concat(m: vec.m16, a: vec.v16, b: vec.v16) -> vec.v16`
+- `vec.v32.concat(m: vec.m32, a: vec.v32, b: vec.v32) -> vec.v32`
+- `vec.v64.concat(m: vec.m64, a: vec.v64, b: vec.v64) -> vec.v64`
+- `vec.v128.concat(m: vec.m128, a: vec.v128, b: vec.v128) -> vec.v128`
+
+
+```python
+def vec.S.concat(m, a, b):
+    begin = -1
+    end = -1
+    for i in range(vec.S.length):
+        if m[i]:
+            end = i + 1
+            if begin < 0:
+                begin = i
+
+    result = vec.S.New()
+    i = 0
+    for j in range(begin, end):
+        result[i] = a[j]
+        i += 1
+    for j in range(0, vec.S.length - i):
+        result[i] = b[j]
+        i += 1
+    return result
+```
+
+### Lane shift
+
+Concats the 2 input vector to form a single double-width vector.
+Shifts this double-width vector by `n` lane to the left (to LSB).
```

Comment:
Shouldn't shifting towards the least significant byte be called a right shift? That is the terminology used by the original proposal.

Reply:
You're absolutely right. I tend to visualize vectors in little-endian, while shifts are most usually viewed as big-endian...